ports/stm32/eth.c

```diff
@@ -805,4 +805,34 @@ int eth_stop(eth_t *self) {
     return 0;
 }
 
+void eth_low_power_mode(eth_t *self, bool enable) {
+    (void)self;
+
+    // Do basic Ethernet initialization here to be able to read/write registers.
+    mp_hal_pin_config_alt_static(MICROPY_HW_ETH_MDC, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, STATIC_AF_ETH_MDC);
+    mp_hal_pin_config_alt_static(MICROPY_HW_ETH_MDIO, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, STATIC_AF_ETH_MDIO);
+    mp_hal_pin_config_alt_static_speed(MICROPY_HW_ETH_RMII_REF_CLK, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_MEDIUM, STATIC_AF_ETH_RMII_REF_CLK);
+
+    // Enable eth clock
+    #if defined(STM32H7)
+    __HAL_RCC_ETH1MAC_CLK_ENABLE();
+    #else
+    __HAL_RCC_ETH_CLK_ENABLE();
+    #endif
```

Comment:
IMO everything up to here (init basic pins, enable peripheral clock) should go in `eth_init()`.  And then the same lines can be removed from `eth_mac_init()` (otherwise there's duplicated code).

Reply:
I agree.

Reply:
Actually the clock still needs to be re-enabled here because it gets disabled in low-power mode.

Reply:
Yes, true.  But at least the pin init can be moved.

Reply:
Okay, one more issue, it will not be possible to put eth in low-power mode from early_init (it draws a lot of power by default) or from powerctrl.c (on stop/standby entry).

Reply:
I don't follow.  Why won't it be possible?  Does it change the current behaviour (which doesn't have low power mode)?

Reply:
@dpgeorge Because the pins won't be initialized if you call eth_low_power_mode() from board_early_init() it won't be able to read/write registers.. same if it's called from `MICROPY_BOARD_ENTER_STOP` before eth is used.

Reply:
I update the PR as requested, please let me know if that's not what you meant or if it needs other changes

Reply:
Can't the board call `eth_init()` before `eth_low_power_mode()`?

From a C API point of view, one should always initialise something before any of its functions are used.

Reply:
Yes I think it's possible, something like:
```
eth_init(&eth_instance, MP_HAL_MAC_ETH0);
eth_low_power_mode(&eth_instance, true);
```